### ISSUE TITLE
Replace macros with proc_macros

### ISSUE THREAD

**Issue by theduke** [graphql-rust/juniper#295]
Replace all macros used for defining GraphQL types with proc macro implementations.

* [x] graphql_interface  (#682, #1009)
* [x] graphql_object  (#181, #333)
* [x] graphql_scalar  (#1014, #1017, #1025)
* [x] graphql_union  (https://github.com/graphql-rust/juniper/commit/ee9a82a8177b50ccf59acaad1e9cce7a73ea9a85, #666)
* [x] ~~graphql_value~~ (not necessary,  #996)

This should probably be a  two-step process: first implement proc-macro alternatives for everything and have tests for functional equivalence.
Then remove the old macros with a new breaking version.

**Comment by samuela** [comment]
Forgive me if this is a noob question, but what's the difference between macros and proc macros?

**Comment by tyranron** [comment]
@samuela 

> Forgive me if this is a noob question, but what's the difference between macros and proc macros?

If it's still relevant, see the official reference:
- [Macros By Example](https://doc.rust-lang.org/reference/macros-by-example.html)
- [Procedural Macros](https://doc.rust-lang.org/reference/procedural-macros.html)

#### TL;DR

Declarative macros:
- declared via pattern-match syntax
- guaranteed to be hygienic
- restricted in expresiveness to certain patterns

Procedural macros:
- declared as a separate crate
- hygiene is up to an implementor
- accept and produce raw tokens, so can generate arbitrary kind of stuff
